README.md
# mabuc

Multi-arm bandits with unobserved confounders

## Usage

`mabuc` reproduces [Bandits with Unobserved Confounders: A Causal Approach](https://proceedings.neurips.cc/paper/2015/hash/795c7a7a5ec6b460ec00c5841019b9e9-Abstract.html) (Elias Bareinboim, Andrew Forney, Judea Pearl, NIPS 2015).

```python
from mabuc.mabuc import Mabuc
m = Mabuc(pr_D=0.5, pr_B=0.5, T=1000)
m.generate_samples(R=1000)
m.get_stats()
```


## Todo

* Display-ad domain 
* Bayesian 
* Offline policy optimization 

## License

`mabuc` was created by Leo Guelman. It is licensed under the terms of the MIT license.



CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.0.1 (21/10/2022)

- First release of `mabuc`!

tests/test_mabuc.py
from mabuc import mabuc


src/mabuc/mabuc.py
"""
Created on Wed Oct  5 07:37:31 2022
@author: lguelman
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

class Mabuc():
    """
    Data Generating Process 
    
    pr_D: probability of a gambler being drunk
    pr_B: probability that machine would be blinking
    pr_payout: probability of payout
    T: number of trials
    """
    
    def __init__(self, pr_D:float, pr_B:float, T:int, pr_payout:float=None):
        self.pr_D = pr_D
        self.pr_B = pr_B
        self.T = T 
        if pr_payout is None:
            pr_payout =  np.array([[0.1, 0.5, 0.4, 0.2],
                                   [0.5, 0.1, 0.2, 0.4]])
        self.pr_payout = pr_payout
        
        
    def sample_players(self):
        
        C = {str(t):None for t in range(self.T)}
        D = np.random.binomial(1, self.pr_D, self.T)
        B = np.random.binomial(1, self.pr_B, self.T)
        
        for t in range(self.T):
            C[str(t)] = {'D':D[t], 'B':B[t]}
 
        self.C = C
         
        return self
    
    def pull_arm(self):
        
        X = {str(t):None for t in range(self.T)}
        
        a0, b0, a1, b1 = (np.ones(4))
        s0, f0, s1, f1 = (np.ones(4))
        
        # seed distribution
        p_Y_X = [self.pr_payout[0][0] * (1- self.pr_D) + self.pr_payout[0][3] * self.pr_D,
                 self.pr_payout[1][1] * (1- self.pr_D) + self.pr_payout[1][2] * self.pr_D]
        
        P_X = [(1-self.pr_D), self.pr_D]
        
        # initialize est. payout for counter-intuition 
        Q1 = [float(np.random.beta(s0, f0, size =1)),  
              float(np.random.beta(s1, f1, size =1))]
                 
           
        for t in range(self.T):

            #---- Play intuition, random, and optimal strategies
            X_int_t = int((self.C.get(str(t))['D'] ^ self.C.get(str(t))['B']))
            X_rnd_t = int(np.random.binomial(1, 0.5, 1))  
            X_opt_t = int(np.array([1, 0, 0, 1])[self.C.get(str(t))['B'] + 2 * self.C.get(str(t))['D']])
            
            #---- Play Thompson Sampling 
            theta0 = float(np.random.beta(a0, b0, size =1))
            theta1 = float(np.random.beta(a1, b1, size =1))
            X_ts_t = np.argmax([theta0, theta1])
            Y_ts_t = self.get_payout_t(X_ts_t, self.C.get(str(t))['D'], self.C.get(str(t))['B'])
           
            if X_ts_t == 0:
                a0+= Y_ts_t
                b0+= 1 - Y_ts_t
            else:
                a1+= Y_ts_t
                b1+= 1 - Y_ts_t
                
            #---- Play Causal Thompson Sampling
            
            # counter-intuition
            if X_int_t == 1:
                X_cint_t = 0
            else:
                X_cint_t=1 
                
            # est. payout for counter-intuition    
            Q1_t = Q1[X_cint_t]
            
            # est. payout for intuition
            Q2_t = p_Y_X[X_int_t]
            w = np.ones(2)
            bias = float(1 - np.abs([Q1_t-Q2_t]))
            if Q1_t > Q2_t:
                w[X_int_t] = bias
            else:
                w[X_cint_t] = bias
            
            X_cts_t = np.argmax([float(np.random.beta(s0, f0, size =1))*w[0],
                                 float(np.random.beta(s1, f1, size =1))*w[1]])
            Y_cts_t = self.get_payout_t(X_cts_t, self.C.get(str(t))['D'], self.C.get(str(t))['B'])
           
            # update counts 
            if X_cts_t == 0:
                s0+= Y_cts_t
                f0+= 1 - Y_cts_t 
               
            else:
                s1+= Y_cts_t
                f1+= 1 - Y_cts_t
               
            # update Q1
            Q1 = [1/P_X[1] * ( s0/(s0+f0) -  p_Y_X[0] * P_X[0] ), 
                  1/P_X[0] * ( s1/(s1+f1) -  p_Y_X[1] * P_X[1] )
                  ]
                
            X[str(t)] = {'X_int':X_int_t, 'X_rnd':X_rnd_t, 'X_opt':X_opt_t, 
                         'X_ts':X_ts_t, 'X_cts':X_cts_t}
         
        self.X = X
    
        return self    
    
    def get_payout_t(self, X, D, B):
        
        if X==0:
            Y = int(np.random.binomial(1, self.pr_payout[0,B + 2*D], 1))
        else: 
            Y = int(np.random.binomial(1, self.pr_payout[1,B + 2*D], 1))
            
        return Y
    
    def get_payout(self):
        
        Y = {str(t):None for t in range(self.T)}
        
        #Y = {'Y'+k[1:5]:dgp.get_payout(v, C.get('D'), C.get('B')) for (k,v) in X.items()}
    
        for t in range(self.T):
            Y[str(t)] = {'Y_int': self.get_payout_t(self.X.get(str(t))['X_int'], self.C.get(str(t))['D'], self.C.get(str(t))['B']),
                         'Y_rnd': self.get_payout_t(self.X.get(str(t))['X_rnd'], self.C.get(str(t))['D'], self.C.get(str(t))['B']),
                         'Y_opt': self.get_payout_t(self.X.get(str(t))['X_opt'], self.C.get(str(t))['D'], self.C.get(str(t))['B']),
                         'Y_ts':  self.get_payout_t(self.X.get(str(t))['X_ts'], self.C.get(str(t))['D'], self.C.get(str(t))['B']),
                         'Y_cts': self.get_payout_t(self.X.get(str(t))['X_cts'], self.C.get(str(t))['D'], self.C.get(str(t))['B']),
                         }
         
        self.Y = Y
        
        return self
    
    def generate_samples(self, R:int=100):
        
        """
        R: Number of repetitions of simulation
        """
        
        data = []
        
        for _ in range(R):
            
            data_r = {str(t):None for t in range(self.T)}
            dgp = Mabuc(pr_D=self.pr_D, pr_B=self.pr_B, T=self.T)
            dgp.sample_players()
            dgp.pull_arm()
            dgp.get_payout()
        
            for t in range(self.T):
                data_r[str(t)] = {**dgp.Y[str(t)],**dgp.X[str(t)], **dgp.C[str(t)]}
         
            df_r = pd.DataFrame.from_dict(data_r, orient='index')
       
            data.append(df_r)
        
        self.data = data 
        
        return self
        
                
    def get_stats(self):
        
        data = self.data
        R = len(data)
       
        int_summary = np.zeros(shape=[R, 2])
        rnd_summary = np.zeros(shape=[R, 2])
        opt_summary = np.zeros(shape=[R, 2])
        ts_summary =  np.zeros(shape=[R, 2])
        cts_summary = np.zeros(shape=[R, 2])
        
        for r in range(R):
            int_summary[r] = data[r].groupby(['X_int'])['Y_int'].agg('mean').values
            rnd_summary[r] = data[r].groupby(['X_rnd'])['Y_rnd'].agg('mean').values
            opt_summary[r] = data[r].groupby(['X_opt'])['Y_opt'].agg('mean').values
            ts_summary[r]  = data[r].groupby(['X_ts'])['Y_ts'].agg('mean').values
            cts_summary[r] = data[r].groupby(['X_cts'])['Y_cts'].agg('mean').values
        
        with np.printoptions(precision=3, suppress=True):
            
            print("\n Optimal Payout [arm1, arm2]: \n")
            print("Mean:", np.mean(opt_summary, axis=0))
            print("S.E.:", np.std(opt_summary, ddof=1, axis=0) / np.sqrt(R))
       
            print("\n Intuition Payout [arm1, arm2]: \n")
            print("Mean:", np.mean(int_summary, axis=0))
            print("S.E.:", np.std(int_summary, ddof=1, axis=0) / np.sqrt(R))
        
            print("\n Random Payout [arm1, arm2]: \n")
            print("Mean:", np.mean(rnd_summary, axis=0))
            print("S.E.:", np.std(rnd_summary, ddof=1, axis=0) / np.sqrt(R))
        
            print("\n Thompson Sampling Payout [arm1, arm2]: \n")
            print("Mean:", np.mean(ts_summary, axis=0))
            print("S.E.:", np.std(ts_summary, ddof=1, axis=0) / np.sqrt(R))
       
            print("\n Causal Thompson Sampling Payout [arm1, arm2]: \n")
            print("Mean:", np.mean(cts_summary, axis=0))
            print("S.E.:", np.std(cts_summary, ddof=1, axis=0) / np.sqrt(R))
            
        self.int_summary = int_summary
        self.rnd_summary = rnd_summary
        self.opt_summary = opt_summary
        self.ts_summary  = ts_summary
        self.cts_summary = cts_summary
       
        return self
        
    
    def plot(self):
        
        data = self.data
        R = len(data)
        
        int_opt = np.zeros(shape=[self.T, R])
        rnd_opt=  np.zeros(shape=[self.T, R])
        ts_opt =  np.zeros(shape=[self.T, R])
        cts_opt = np.zeros(shape=[self.T, R])
        
        # played optimal action?
        for r in range(R):
            int_opt[:,r] = data[r]['X_int'] == data[r]['X_opt']
            rnd_opt[:,r] = data[r]['X_rnd'] == data[r]['X_opt']
            ts_opt[:,r] = data[r]['X_ts'] == data[r]['X_opt']
            cts_opt[:,r] = data[r]['X_cts'] == data[r]['X_opt']
         
        int_opt_summary = np.mean(int_opt, axis=1)
        rnd_opt_summary = np.mean(rnd_opt, axis=1)
        ts_opt_summary = np.mean(ts_opt, axis=1)
        cts_opt_summary = np.mean(cts_opt, axis=1)
        
        cumsum_ones = np.cumsum(np.ones(self.T))
        
        plt.plot(cumsum_ones, int_opt_summary, label ="Intuition")
        plt.plot(cumsum_ones, rnd_opt_summary, label ="Random")
        plt.plot(cumsum_ones, ts_opt_summary, label ="TS")
        plt.plot(cumsum_ones, cts_opt_summary, label ="CTS")
        plt.title('Probability of Optimal Action')
        plt.xlabel('Trial')
        plt.ylabel('Probability')
        plt.legend()
        plt.ylim((0,1))
        plt.show()
        
        



src/mabuc/Blbfuc.py
"""
Created on Sat Oct 22 17:31 2022
@author: lguelman
"""

import numpy as np
import pandas as pd
import matplotlib.pyplot as plt

class Blbfuc():
    """
    Batch Learning from Bandit Feedback with Unobserved Confounders
    
    pr_I: probability of high-income
    pr_M: probability of male
    pr_payout: probability of payout
    T: number of observations
    """
    
    def __init__(self, pr_I:float, pr_M:float, T:int, pr_payout:float=None):
        self.pr_I = pr_I
        self.pr_M = pr_M
        self.T = T 
        if pr_payout is None:
            pr_payout =  np.array([[0.25, 0.5, 0.4, 0.2],
                                   [0.5, 0.25, 0.2, 0.4]])
        self.pr_payout = pr_payout
        
        
    def sample_clients(self):
        
        C = {str(t):None for t in range(self.T)}
        I = np.random.binomial(1, self.pr_I, self.T)
        M = np.random.binomial(1, self.pr_M, self.T)
        
        for t in range(self.T):
            C[str(t)] = {'I':I[t], 'M':M[t]}
 
        self.C = C
         
        return self
    
    def play_policy(self):
        
        X = {str(t):None for t in range(self.T)}
       
        #---- Play client choice (clc), random (rct), and oracle strategies
        for t in range(self.T):
            X_clc_t = int((self.C.get(str(t))['I'] ^ self.C.get(str(t))['M']))
            # I==0 and M==0 | I==1 and M==1 ==> X=0
            # I==0 and M==1 | I==1 and M==0 == >X=1
            X_rct_t = int(np.random.binomial(1, 0.5, 1))  
            X_ora_t = int(np.array([1, 0, 0, 1])[self.C.get(str(t))['M'] + 2 * self.C.get(str(t))['I']])
            # I==0 and M==0 ==> X=1
            # I==0 and M==1 ==> X=0
            # I==1 and M==0 ==> X=0
            # I==1 and M==1 ==> X=1
            X_clc_opt_t = int(self.C.get(str(t))['M'])
            # M==0 => X==0
            # M==1 => X==1
            X_rct_opt_t = 1 - int(self.C.get(str(t))['M'])
            # M==0 ==> X==1
            # M==1 ==> X==0
            
            X[str(t)] = {'X_clc':X_clc_t, 'X_rct':X_rct_t, 'X_ora':X_ora_t,
                         'X_clc_opt':X_clc_opt_t, 'X_rct_opt':X_rct_opt_t}
  
        self.X = X

        return self    
    
    
    def get_payout_t(self, X, I, M):
        
        if X==0:
            Y = int(np.random.binomial(1, self.pr_payout[0,M + 2*I], 1))
        else: 
            Y = int(np.random.binomial(1, self.pr_payout[1,M + 2*I], 1))
            
        return Y
    
    #X==0
    #I==0 and M==0 ==> 0.25
    #I==0 and M==1 ==> 0.50
    #I==1 and M==0 ==> 0.40
    #I==1 and M==1 ==> 0.20
           
    
    def get_payout(self):
        
        Y = {str(t):None for t in range(self.T)}
        
        #Y = {'Y'+k[1:5]:dgp.get_payout(v, C.get('D'), C.get('B')) for (k,v) in X.items()}
    
        for t in range(self.T):
            Y[str(t)] = {'Y_clc': self.get_payout_t(self.X.get(str(t))['X_clc'], self.C.get(str(t))['I'], self.C.get(str(t))['M']),
                         'Y_rct': self.get_payout_t(self.X.get(str(t))['X_rct'], self.C.get(str(t))['I'], self.C.get(str(t))['M']),
                         'Y_ora': self.get_payout_t(self.X.get(str(t))['X_ora'], self.C.get(str(t))['I'], self.C.get(str(t))['M']),
                         'Y_clc_opt': self.get_payout_t(self.X.get(str(t))['X_clc_opt'], self.C.get(str(t))['I'], self.C.get(str(t))['M']),
                         'Y_rct_opt': self.get_payout_t(self.X.get(str(t))['X_rct_opt'], self.C.get(str(t))['I'], self.C.get(str(t))['M']),
                         }
         
        self.Y = Y
        
        return self
    
    def _regret_decision_criterion(self, data):
    
        R = len(data)
        
        for r in range(R):
            
            Py_dox0_m0, Py_dox0_m1, Py_dox1_m0, Py_dox1_m1 =  \
                data[r].groupby(['X_rct', 'M'])['Y_rct'].agg('mean').values
            
            Py_x0_m0, Py_x0_m1, Py_x1_m0, Py_x1_m1 =  \
                data[r].groupby(['X_clc', 'M'])['Y_clc'].agg('mean').values
                
            Px1_m0, Px1_m1  = \
                data[r].groupby(['M'])['X_clc'].agg('mean').values
            
            Px0_m0 = 1-Px1_m0
            Px0_m1 = 1-Px1_m1
            
            Pyx1_x0_m1 = \
                (Py_dox1_m1 - Py_x1_m1 * Px1_m1) * 1/Px0_m1
                
            Pyx1_x0_m0 = \
                (Py_dox1_m0 - Py_x1_m0 * Px1_m0) * 1/Px0_m0
                  
            Pyx0_x1_m1 = \
                (Py_dox0_m1 - Py_x0_m1 * Px0_m1) * 1/Px1_m1
                
            Pyx0_x1_m0 = \
                (Py_dox0_m0 - Py_x0_m0 * Px0_m0) * 1/Px1_m0
                
            M1_X_clc0 = Pyx1_x0_m1 - Py_x0_m1 > 0
            M1_X_clc1 = Py_x1_m1 - Pyx0_x1_m1 > 0
            M0_X_clc0 = Pyx1_x0_m0 - Py_x0_m0 > 0
            M0_X_clc1 = Py_x1_m0 - Pyx0_x1_m0 > 0
            
            
            def _action_map(df, M1_X_clc0, M1_X_clc1, M0_X_clc0, M0_X_clc1):
            
                M = df['M']
                X_clc = df['X_clc']
            
                if M==1 and X_clc==0:
                    if M1_X_clc0: 
                        X_rdc = 1
                    else:
                        X_rdc = 0
                    
                elif M==1 and X_clc==1:
                    if M1_X_clc1:
                        X_rdc = 1
                    else:
                        X_rdc = 0
                
                elif M==0 and X_clc==0:
                    if M0_X_clc0:
                        X_rdc = 1
                    else:
                        X_rdc = 0
                
                elif M==0 and X_clc==1:
                    if M0_X_clc1:
                        X_rdc = 1
                    else:
                        X_rdc = 0
                        
                return  X_rdc 
                    
           
            data[r]['X_rdc'] = data[r].apply(_action_map, 
                                             args=(M1_X_clc0, M1_X_clc1, M0_X_clc0, M0_X_clc1), axis=1)   
            
            Y_rdc = []
            for t in range(self.T):
                Y_rdc.append(self.get_payout_t(X=data[r]['X_rdc'].iloc[t], 
                                  I = data[r]['I'].iloc[t],
                                  M = data[r]['M'].iloc[t]))
                
            data[r]['Y_rdc'] = Y_rdc
            
        return data
            
    
    def generate_samples(self, R:int=100):
        
        """
        R: Number of repetitions of simulation
        """
        
        data = []
        
        for _ in range(R):
            
            data_r = {str(t):None for t in range(self.T)}
            dgp = Blbfuc(pr_I=self.pr_I, pr_M=self.pr_M, T=self.T)
            dgp.sample_clients()
            dgp.play_policy()
            dgp.get_payout()
        
            for t in range(self.T):
                data_r[str(t)] = {**dgp.Y[str(t)],**dgp.X[str(t)], **dgp.C[str(t)]}
         
            df_r = pd.DataFrame.from_dict(data_r, orient='index')
       
            data.append(df_r)
        
        data = self._regret_decision_criterion(data)
        
        self.data = data 
        
        return self

    
    def get_stats(self):
        
        data = self.data
        R = len(data)
       
        clc_summary = np.zeros(shape=[R, 4])
        rct_summary = np.zeros(shape=[R, 4])
        ora_summary = np.zeros(shape=[R, 4])
        #rdc_summary = np.zeros(shape=[R, 4])
      
        
        for r in range(R):
            clc_summary[r] = data[r].groupby(['X_clc', 'M'])['Y_clc'].agg('mean').values
            rct_summary[r] = data[r].groupby(['X_rct', 'M'])['Y_rct'].agg('mean').values
            ora_summary[r] = data[r].groupby(['X_ora', 'M'])['Y_ora'].agg('mean').values
            #rdc_summary[r] = data[r].groupby(['X_rdc', 'M'])['Y_rdc'].agg('mean').values
            
        
        with np.printoptions(precision=3, suppress=True):
            
            print("\n Oracle Payout [P-F, P-M, D-F, D-M]: \n")
            print("Mean:", np.mean(ora_summary, axis=0))
            print("S.E.:", np.std(ora_summary, ddof=1, axis=0) / np.sqrt(R))
       
            print("\n Client Choice Payout [P-F, P-M, D-F, D-M]: \n")
            print("Mean:", np.mean(clc_summary, axis=0))
            print("S.E.:", np.std(clc_summary, ddof=1, axis=0) / np.sqrt(R))
        
            print("\n RCT Payout [P-F, P-M, D-F, D-M]: \n")
            print("Mean:", np.mean(rct_summary, axis=0))
            print("S.E.:", np.std(rct_summary, ddof=1, axis=0) / np.sqrt(R))
            
            #print("\n RCT Payout [P-F, P-M, D-F, D-M]: \n")
            #print("Mean:", np.mean(csl_summary, axis=0))
            #print("S.E.:", np.std(csl_summary, ddof=1, axis=0) / np.sqrt(R))
            
        
        ymean_ora = np.zeros(shape=[R])
        ymean_clc = np.zeros(shape=[R])
        ymean_rct = np.zeros(shape=[R])
        ymean_clc_opt = np.zeros(shape=[R])
        ymean_rct_opt = np.zeros(shape=[R])
        ymean_rdc_opt = np.zeros(shape=[R])
        
        for r in range(R):
            ymean_ora[r] = data[r]['Y_ora'].agg('mean')
            ymean_clc[r] = data[r]['Y_clc'].agg('mean')
            ymean_rct[r] = data[r]['Y_rct'].agg('mean')
            ymean_clc_opt[r] = data[r]['Y_clc_opt'].agg('mean')
            ymean_rct_opt[r] = data[r]['Y_rct_opt'].agg('mean')
            ymean_rdc_opt[r] = data[r]['Y_rdc'].agg('mean')
        
        
        with np.printoptions(precision=3, suppress=True):
            
            print("\n Oracle Payout \n")
            print("Mean:", np.mean(ymean_ora))
            print("S.E.:", np.std(ymean_ora, ddof=1) / np.sqrt(R))
            
            print("\n Client choice Payout \n")
            print("Mean:", np.mean(ymean_clc))
            print("S.E.:", np.std(ymean_clc, ddof=1) / np.sqrt(R))
            
            print("\n RCT Payout \n")
            print("Mean:", np.mean(ymean_rct))
            print("S.E.:", np.std(ymean_rct, ddof=1) / np.sqrt(R))
            
            print("\n Client choice Optimal Payout \n")
            print("Mean:", np.mean(ymean_clc_opt))
            print("S.E.:", np.std(ymean_clc_opt, ddof=1) / np.sqrt(R))
            
            print("\n RCT Optimal Payout \n")
            print("Mean:", np.mean(ymean_rct_opt))
            print("S.E.:", np.std(ymean_rct_opt, ddof=1) / np.sqrt(R))
       
            print("\n RDC Optimal Payout \n")
            print("Mean:", np.mean(ymean_rdc_opt))
            print("S.E.:", np.std(ymean_rdc_opt, ddof=1) / np.sqrt(R))
   
        
        self.clc_summary = clc_summary
        self.rct_summary = rct_summary
        self.ora_summary = ora_summary
        self.ymean_ora = ymean_ora
        self.ymean_clc = ymean_clc
        self.ymean_rct = ymean_rct
        self.ymean_rdc_opt = ymean_rdc_opt
        self.ymean_rdc_opt = ymean_rdc_opt
        
        return self
        
    
    
        
        
###

m = Blbfuc(pr_I=0.5, pr_M=0.5, T=1000)
m.generate_samples(R=100)
m.get_stats()

        

